File: src/darkfactory/utils/shell.py
"""Shell command runner shared by workflow and system runners."""

from __future__ import annotations

import os
import subprocess
from pathlib import Path


def run_shell(
    cmd: str,
    cwd: Path,
    env: dict[str, str],
) -> subprocess.CompletedProcess[str]:
    """Run a shell command once and return the completed-process result."""
    full_env = dict(os.environ)
    full_env.update(env)

    return subprocess.run(
        cmd,
        shell=True,
        cwd=str(cwd),
        capture_output=True,
        text=True,
        env=full_env,
        check=False,
    )


def run_foreground(cmd: list[str], *, cwd: Path | None = None) -> int:
    """Run a command with stdout/stderr flowing to the terminal.

    Returns the process exit code. No output is captured.
    """
    result = subprocess.run(cmd, cwd=str(cwd) if cwd else None, check=False)
    return result.returncode

File: ts/src/utils/shell.ts
/**
 * shell.ts — shell command execution helpers.
 *
 * Both functions delegate to utils/subprocess.ts.
 */

import { type Result, err, ok } from "./result.js";
import {
  type ExecErr,
  type ExecResult,
  execForeground,
  execShell,
} from "./subprocess.js";

export type { ExecResult };

/**
 * Run shell command string, capture output.
 */
export async function runShell(
  cmd: string,
  cwd: string,
  env?: Record<string, string>
): Promise<ExecResult> {
  return execShell(cmd, env !== undefined ? { cwd, env } : { cwd });
}

/**
 * Run command in foreground, stream to terminal.
 * Returns exit code on success, ExecErr on spawn failure.
 */
export async function runForeground(
  cmd: readonly string[],
  cwd?: string
): Promise<Result<number, ExecErr>> {
  try {
    const exitCode = await execForeground(cmd, cwd);
    return ok(exitCode);
  } catch (e) {
    const execErr: ExecErr = {
      kind: "exec-err",
      exitCode: -1,
      stderr: String(e),
      cmd,
    };
    return err(execErr);
  }
}
